### PR TITLE
Fixed ucwords to be functionally the same as in-built PHP version

### PR DESCRIPTION
**Pull request by martinkiva**

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
As you can see, this breaks the tests with TurkishUtf8.
ucfirst and lcfirst need no patch at all, they both have the behavior you expected them to have.
For ucwords, "space" is not the only character that should trigger capitalization.
You need a regexp to catch all first chars of "words". I suggest `/\b(.)/` and applying mb_convert_case on the match

**Comment by martinkiva** [comment]
My bad, I didn't run the TurkishUtf8Test locally. Good point on the word boundaries. I've updated ucwords to use the regex you provided. The unit tests all pass locally now.

The reason I changed ucfirst and lcfirst is that previously they used ucwords and I wanted to ucwords to use ucfirst. My latest commits doesn't require that anymore, so I've changed them back. However, ucfirst and lcfirst both used non multibyte functions (substr and strlen). Is that correct?

One other tangential thing, in the ucfirst and lcfirst functions, you use iconv_substr. This function is about 60% slower than mb_substr. Can they be changed over?

File: class/Patchwork/Utf8.php [hunk 1]
```diff
@@ -555,7 +555,7 @@ static function substr_replace($s, $replace, $start, $len = 2147483647)
     static function ucfirst($s)
     {
         $c = iconv_substr($s, 0, 1, 'UTF-8');
-        return static::ucwords($c) . substr($s, strlen($c));
+        return static::strtoupper($c) . substr($s, strlen($c));
```
Comment (nicolas-grekas) [inline class/Patchwork/Utf8.php]: Why? Title case is not the same as upper case, and here, we need title case (see http://userguide.icu-project.org/transforms/casemappings)

File: class/Patchwork/Utf8.php [hunk 2]
```diff
@@ -566,7 +566,20 @@ static function lcfirst($s)
 
     static function ucwords($s)
     {
-        return mb_convert_case($s, MB_CASE_TITLE, 'UTF-8');
+        if (PHP_VERSION_ID < 50401)
+        {
+            return mb_ereg_replace("\b(.)", "mb_convert_case('\\1', MB_CASE_UPPER, 'UTF-8')", $s, 'e');
```
Comment (nicolas-grekas) [inline class/Patchwork/Utf8.php]: preg_replace please + MB_CASE_TITLE

**Comment by nicolas-grekas** [comment]
Yes, using strlen+substr is on purpose, because in this case, we don't need multi-byte handlig but for the first utf-8 char.
Are you sure about iconv_strlen being faster than mb_strlen? I bet this is ot true on every platform (Windows, Linux+libc, other). Do you have any facts?

File: class/Patchwork/Utf8.php (hunk 1 above)
Comment (martinkiva) [inline class/Patchwork/Utf8.php]: We are only converting one character here, so shouldn't upper case be the same as title case? strtoupper should be faster than ucwords for a single character. Also, this now mirrors lcwords, which uses strtolower.

Comment (nicolas-grekas) [inline class/Patchwork/Utf8.php]: You didn't read the ICU page :) The DZ ligature e.g. exists as a single UTF-8 char.
This one has 3 char cases:
- lower, which is a single `dz` char
- upper, which is a single `DZ` char
- title, which is a single `Dz` char

As you see, lcfirst must use strtolower, and ucfirst the title case version

File: class/Patchwork/Utf8.php (hunk 2 above)
Comment (martinkiva) [inline class/Patchwork/Utf8.php]: I'll change it to preg_replace_callback. Again the MB_CASE_TITLE vs MB_CASE_UPPER shouldn't matter as it is one character. Also, The original bug here is that ucwords is not supposed to change the case to title case. It is just supposed to change the first character of each word to upper case.

File: class/Patchwork/Utf8.php (hunk 1 above)
Comment (martinkiva) [inline class/Patchwork/Utf8.php]: I see! I'll make the change.

File: class/Patchwork/Utf8.php (hunk 2 above)
Comment (nicolas-grekas) [inline class/Patchwork/Utf8.php]: Again, TITLE vs UPPER *IS* different for single characters.

**Comment by nicolas-grekas** [comment]
Thanks you @martinkiva

**Pull request merged**